### DISCUSSION TITLE
The subtle semantic difference between the 'get' and 'set' methods

### DISCUSSION THREAD

**Discussion by e7h4n** [pmndrs/jotai#2793] [General]
1. It seems that `get` should be idempotent. In a well-implemented jotai system, repeatedly calling `get` on atoms should yield the same final state of the entire store as calling `get` just once. In contrast, `set` does not have this property.

2. The computation within `get` is lazy. In a dependency chain formed by `get`, when upstream atom properties change, the computations encapsulated in all `get` calls will not execute immediately; instead, they will be delayed until the downstream atom is accessed via `get`. Conversely, `set` actively pushes the new values to all affected atoms.

3. The differences in execution between `get` and `set` feel somewhat similar to the distinction between cold and hot observables in Rx.

**Comment by e7h4n** [comment]
In actual application, the scenarios for `set` are more extensive, so some colleagues suggest using `invoke` to replace `set`

**Comment by dai-shi** [comment]
> 2\. The computation within `get` is lazy.

Well, it depends. If the atom is mounted, it's re-computed eagerly.
